src/snowflake/cli/_plugins/nativeapp/entities/application_package.py

```diff
@@ -543,9 +605,41 @@ def action_version_drop(
             f"Version {version} in application package {self.name} dropped successfully."
         )
 
-    def _bundle(self):
+    def _bundle(self, action_ctx: ActionContext = None):
         model = self._entity_model
         bundle_map = build_bundle(self.project_root, self.deploy_root, model.artifacts)
+        if self._entity_model.children:
+            # Create _entities directory
+            entities_dir = self.entities_deploy_root
+            os.makedirs(entities_dir)
+            children_sql = []
+            for child in self._entity_model.children:
+                # Create child sub directory
+                child_artifacts_dir = entities_dir / child.target
+                os.makedirs(child_artifacts_dir)
+                child_entity = action_ctx.get_entity(child.target)
+                child_entity.bundle(child_artifacts_dir)
+                app_role = child.application_role
+                child_schema = child.identifier.schema_
+                children_sql.append(
+                    child_entity.get_deploy_sql(
+                        schema=child_schema,
+                        from_=Path(self._entity_model.entities_dir, child.target),
+                    )
+                )
+                if app_role:
+                    children_sql.append(
+                        f"CREATE APPLICATION ROLE IF NOT EXISTS {app_role};"
+                    )
+                    if child_schema:
+                        children_sql.append(
+                            f"GRANT USAGE ON SCHEMA {child_schema} TO APPLICATION ROLE {app_role};"
+                        )
+                if isinstance(child_entity, StreamlitEntity) and app_role:
```

Comment:
Let's think of ways to preserve encapsulation in the future? IMHO it should be possible to add new entities that can be children of an app without modifying the native app code at all (e.g. if SPCS wants to enable the feature on their own, it should be self-contained)

Reply:
Added `get_grants_sql()` to the required interface each child has to implement. WDYT?